### PR TITLE
[zero] adapt zero for unsharded parameters

### PR DESCRIPTION
**Pull request by 1SAA**
* support existing sharded and unsharded parameters in zero

* add unitest for moe-zero model init

### CODE REVIEW

File: colossalai/engine/ophooks/__init__.py [hunk 1]
```diff
@@ -73,6 +73,10 @@ def register_ophooks_recursively(module: torch.nn.Module, ophook_list: List[Base
     if len(list(module.parameters(recurse=False))) == 0:
         return
 
+    # Unsharded module do not need hooks
+    if not module.param_is_sharded:
```
Comment (feifeibear) [inline colossalai/engine/ophooks/__init__.py:77]: Wrong. This is a common utils. We can not assume it is only used for shared tensors...

File: colossalai/nn/layer/moe/utils.py [hunk 2]
```diff
@@ -51,6 +52,12 @@ def __call__(self, inputs: torch.Tensor):
         return inputs * noisy
 
 
+def autocast_softmax(logit: torch.Tensor, dim: int):
+    if logit.dtype != torch.float32:
+        logit = logit.float()
+    return F.softmax(logit, dim=dim)
```
Comment (feifeibear) [inline colossalai/nn/layer/moe/utils.py:58]: return F.softmax(logit.float() if logit.dtype != torch.float32 else logit, dim=dim)

File: colossalai/zero/init_ctx/init_context.py [hunk 3]
```diff
@@ -88,6 +88,8 @@ class ZeroContextConfig(object):
     """The configuration used to control zero context initialization.
 
     Args:
+        is_replicated (bool, optional): Is param replicated across data parallel group.
```
Comment (feifeibear) [inline colossalai/zero/init_ctx/init_context.py:91]: raning to use_replicated.
Do not use is_xxx as args

File: colossalai/zero/sharded_model/sharded_model_v2.py [hunk 4]
```diff
@@ -245,17 +272,27 @@ def _post_backward_operations(self) -> None:
...
-            if self.reuse_fp16_shard:
-                grad_fp16_payload = p.colo_attr.sharded_data_tensor.payload
+            if p.colo_attr.param_is_sharded:
+                if self.reuse_fp16_shard:
+                    grad_fp16_payload = p.colo_attr.sharded_data_tensor.payload
+                else:
+                    grad_fp16_payload = p.colo_attr.fp16_grad.payload
             else:
-                grad_fp16_payload = cast_tensor_to_fp32(p.colo_attr.fp16_grad.payload)
+                grad_fp16_payload = p.grad.data
```
Comment (feifeibear) [inline colossalai/zero/sharded_model/sharded_model_v2.py:286]: why change the original logic...

File: tests/test_moe/test_moe_zero_model.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    # check whether parameters are identical in ddp
+    for name, p in zero_model.named_parameters():
+        if not p.colo_attr.param_is_sharded and p.is_replicated:
+            assert_equal_in_group(p.data)
+
+    model = MoeModel().half()
+    col_model_deepcopy(zero_model, model)
+    model = model.cuda()
+    grad_handler = MoeGradientHandler(model, 0)    # 0 is dummy parameter
```
Comment (feifeibear) [inline tests/test_moe/test_moe_zero_model.py:48]: why not override the __init__ for MoeGradientHander.
You won't used self.optimizer for MOE?

File: colossalai/zero/sharded_model/sharded_model_v2.py [hunk 6]
```diff
@@ -178,14 +189,15 @@ def forward(self, *args: Any, **kwargs: Any) -> torch.Tensor:
         return outputs
 
     def backward(self, loss):
+        # self._prev_backward_operations()
```
Comment (feifeibear) [inline colossalai/zero/sharded_model/sharded_model_v2.py:192]: remove it

File: colossalai/nn/layer/moe/utils.py (hunk 2 above)
Comment (1SAA) [reply colossalai/nn/layer/moe/utils.py:58]: What's the necessity to write like that?

File: colossalai/zero/init_ctx/init_context.py (hunk 3 above)
Comment (1SAA) [reply colossalai/zero/init_ctx/init_context.py:91]: We do not use the parameters as replicated. 
But the parameters are replicated inherently.

Comment (1SAA) [reply colossalai/zero/init_ctx/init_context.py:91]: Maybe just use 'replicated' instead.

File: colossalai/zero/sharded_model/sharded_model_v2.py (hunk 4 above)
Comment (1SAA) [reply colossalai/zero/sharded_model/sharded_model_v2.py:286]: Since github mismatched changed lines.

**Comment by ver217** [comment]
Hi, CodeFactor reports 1 issue. Could you fix it?

File: colossalai/engine/ophooks/__init__.py (hunk 1 above)
Comment (1SAA) [reply colossalai/engine/ophooks/__init__.py]: done

File: tests/test_moe/test_moe_zero_model.py (hunk 5 above)
Comment (1SAA) [reply tests/test_moe/test_moe_zero_model.py]: done

File: colossalai/zero/sharded_model/sharded_model_v2.py (hunk 6 above)
Comment (1SAA) [reply colossalai/zero/sharded_model/sharded_model_v2.py]: done

**Comment by 1SAA** [comment]
> Hi, CodeFactor reports 1 issue. Could you fix it?

I think it will get removed when PR 571 merged into main.

File: colossalai/zero/sharded_model/sharded_model_v2.py [hunk 7]
```diff
@@ -202,6 +212,17 @@ def _update_memstats(self):
             self._cuda_margin_space = colo_cuda_memory_capacity() - max(self._memstats_collector.overall_cuda)
         self._iter_cnter += 1
 
+    @torch.no_grad()
+    def _prev_backward_operations(self) -> None:
```
Comment (feifeibear) [inline colossalai/zero/sharded_model/sharded_model_v2.py:216]: did you use it?

File: colossalai/nn/layer/moe/utils.py (hunk 2 above)
Comment (feifeibear) [reply colossalai/nn/layer/moe/utils.py:58]: That's OK

**Review by feifeibear — APPROVED** [review]

File: colossalai/zero/sharded_model/sharded_model_v2.py (hunk 7 above)
Comment (1SAA) [reply colossalai/zero/sharded_model/sharded_model_v2.py]: Removed after PR 571.
